agent/consul/rpc.go

```diff
@@ -968,6 +968,32 @@ RUN_QUERY:
 
 	// Execute the queryFn
 	err := fn(ws, state)
+
+	// If the request is unauthenticated, we blank out ResultsFilteredByACLs to
+	// limit information leaking.
+	//
+	// Endpoints that support bexpr filtering could be used in combination with
+	// this flag/header to discover the existence of resources to which the user
+	// does not have access, therefore we only expose it when the user presents
+	// a valid ACL token. This doesn't completely remove the risk (by nature the
+	// purpose of this flag is to let the user know there are resources they can
+	// not access) but it prevents completely unauthenticated users from doing so.
+	//
+	// Notes:
+	//
+	//	* The definition of "unauthenticated" here is incomplete, as it doesn't
+	//	  account for the fact that operators can modify the anonymous token with
+	//	  custom policies, or set namespace default policies. As these scenarios
+	//	  are less common and this flag is a best-effort UX improvement, we think
+	//	  the trade-off for reduced complexity is acceptable.
+	//
+	//	* It's safe to assume that the token has already been validated (and just
+	//	  check that it's not blank) at this point, because ResultsFilteredByACLs
+	//	  is only set to true when applying the already-resolved token's policies.
+	if queryOpts.GetToken() == "" {
+		queryMeta.SetResultsFilteredByACLs(false)
+	}
```

Comment:
This `blockingQuery` function does handle most "list" RPCs, however I think there are a few special cases that don't use `blockingQuery`, where we'll want to apply this same logic. Streaming is one case, `Txn.Read`, `PreparedQuery.Execute`, and `PreparedQuery.Explain` are a few more. I found some of these by looking at callers to `Server.setQueryMeta`.

Generally I also try to avoid large inline comments in functions, as it makes the code harder to read (although definitely the Consul code does not follow this convention well). Instead of a large inline comment I find a separate function with a godoc string on it works better as it makes the scope more explicit.

I think both of these would be addressed by making a new function for this operation, and moving the comment to be a godoc string for that function, and calling it here.

Reply:
Good call! I've extracted this into a method that is called as part of `setQueryMeta`, so those special cases should be covered.

As for streaming, it looks like ACL-filtering is done as part of the `Subscribe` endpoint:
https://github.com/hashicorp/consul/blob/3666401ae3ae924b1d1acd9261f5e187c610ac07/agent/rpc/subscribe/subscribe.go#L78-L80

And the event payload doesn't seem to have an embedded `QueryMeta` struct, so I'm not 100% on how it should work, What do you think of deferring this for now?

Reply:
Ya, good call. It's probably less relevant for streaming.